mapie/control_risk/ltt.py

```diff
@@ -0,0 +1,103 @@
+import numpy as np
+import warnings
+from numpy.typing import NDArray
+from typing import Tuple, List, Union, Optional, Any
+from .p_values import hoefdding_bentkus_p_value
+
+
+def _ltt_procedure(
+    r_hat: NDArray,
+    alpha_np: NDArray,
+    delta: Optional[float],
+    n_obs: int
+) -> Tuple[List[List[Any]], NDArray]:
+    """
+    Apply the ltt procedure for risk control
+    should be precision for multilabelclassif
+    Note that we will do a multipletest for r_hat that are
+    less than level alpha.
+    Procedure:
+        - compute p_values for each lambdas descretize
+        - Apply a fwer algorithm, here Bonferonni correction
+        - Return the index lambdas that give you the control
+        at alpha level
+        - n_obs is the length of data cal
+    Parameters
+    ----------
+    r_hat: NDArray of shape (n_samples, )
+        Empirical risk of metric_control with respect
+        to the lambdas.
+    alpha: NDArray of alphas level.
+    delta: Float value
+        Correspond to proportion of failure we don't
+        want to exceed.
+    Returns
+    ----------
+    valid_index: NDArray of shape (n_alpha, ).
+        Contain the valid index that satisfy fwer control
+        for each alpha (shape aren't the same for each alpha)
+    p_values: NDArray of shape (n_lambda, n_alpha)
+        Contains the values of p_value for different alpha
+    """
+    if delta is None:
+        raise ValueError(
+            "Invalid delta: delta cannot be None while"
+            + " using LTT for precision control. "
+        )
+    p_values = hoefdding_bentkus_p_value(r_hat, n_obs, alpha_np)
+    valid_index = []
+    for i in range(len(alpha_np)):
+        N_coarse = len(np.where(p_values[:, i] < delta/n_obs)[0])
+        if N_coarse == 0:
+            l_index = []  # type: List[int]
+            valid_index.append(l_index)
+        else:
+            l_index = np.where(p_values[:, i] <= delta/n_obs)[0].tolist()
+            valid_index.append(l_index)
```

Comment:
ENH: remove indent (because it is common to `if N_coarse == 0:` and `else:` conditions) => remove also l53.

Reply:
Done.

Reply:
Done